src/mesh/exodusII_io_helper.C

```diff
@@ -1730,18 +1731,21 @@ void ExodusII_IO_Helper::initialize_element_variables(std::vector<std::string> n
 
   this->write_var_names(ELEMENTAL, names);
 
-  // Form the element variable truth table and send to Exodus.
-  // This tells which variables are written to which blocks,
-  // and can dramatically speed up writing element variables
-  //
-  // We really should initialize all entries in the truth table to 0
-  // and then loop over all subdomains, setting their entries to 1
-  // if a given variable exists on that subdomain.  However,
-  // we don't have that information, and the element variables
-  // passed to us are padded with zeroes for the blocks where
-  // they aren't defined.  To be consistent with that, fill
-  // the truth table with ones.
-  std::vector<int> truth_tab(num_elem_blk*num_elem_vars, 1);
+  // Use the truth table to indicate which subdomain/variable pairs are
+  // active according to vars_active_subdomains.
+  std::vector<int> truth_tab(num_elem_blk*num_elem_vars, 0);
+  for(unsigned int var_num=0; var_num<vars_active_subdomains.size(); var_num++)
+    {
+      for(auto block_id : vars_active_subdomains[var_num])
+        {
+          unsigned int block_index =
+            std::distance(block_ids.begin(), std::find(block_ids.begin(), block_ids.end(), block_id));
+
+          unsigned int truth_tab_index = block_index*num_elem_vars + var_num;
+          truth_tab[truth_tab_index] = 1;
+        }
+    }
+
```

Comment:
Is this the backwards-incompatible change that is causing MOOSE tests to fail? We are now defaulting to all zeros in the `truth_tab`?

Reply:
We currently default to "all ones" in the truth table, i.e. write out element data for all blocks.

This change should only affect cases where you have CONSTANT MONOMIAL variables which are not active on all subdomains, since otherwise the truth table should be "all ones" still, just like before. Is that the case in the MOOSE tests that are failing?

Reply:
Both tests which are failing are attempting to read in Exodus files written out by other tests. Both tests have CONSTANT, MONOMIAL auxiliary variables in them, but they aren't subdomain restricted as far as I can tell. Also the first parts of each test still pass, so presumably the outputs from those are identical to before?

Reply:
Could something be going wrong with the writing?  I grabbed a copy of the mesh from the failing-in-Civet MOOSE test, and I'm reading it and copying a solution variable from it in a new unit test, and it's working fine.

I'll try rebuilding MOOSE and reproducing the failure and see if *that* output breaks the unit test.

Reply:
Yeah, something's got to be an issue with the writing or with the reading in of this branch's output.  tests/restart/restart/elem_part1_out.e went from 86100 bytes to 66840 bytes.

And there aren't any subdomains in that MOOSE test!

Reply:
OK, sounds like we're that we're closing in on the issue. Can you compare the ncdump of the two files (or send them to me so I can do it)? Perhaps it's not writing out some of the element variables now?

Reply:
@roystgnr do you have any further info on this? Is there anything that I can run to reproduce the issue?

Reply:
Sorry for the delay; I've been stuck on other work after getting what I *thought* should be a decent unit test case set up.  But my unit test doesn't even work on master:

```
  void testExodusCopyElementSolution () 
  {  
    {  
      Mesh mesh(*TestCommWorld); 
 
      EquationSystems es(mesh); 
      System &sys = es.add_system<System> ("SimpleSystem"); 
      sys.add_variable("e", CONSTANT, MONOMIAL); 
 
      MeshTools::Generation::build_square (mesh, 
                                           3, 3, 
                                           0., 1., 0., 1.); 
 
      es.init(); 
 
      ExodusII_IO exii(mesh); 
 
      exii.write_equation_systems("mesh_with_soln.e", es); 
    } 
 
    // copy_elemental_solution currently requires ReplicatedMesh 
    {  
      ReplicatedMesh mesh(*TestCommWorld); 
 
      EquationSystems es(mesh); 
      System &sys = es.add_system<System> ("SimpleSystem"); 
      sys.add_variable("teste", CONSTANT, MONOMIAL); 
 
      ExodusII_IO exii(mesh); 
 
      exii.read("mesh_with_soln.e"); 
      mesh.prepare_for_use(); 
 
      es.init(); 
 
      // Read the solution e on time step 6 into variable teste 
      exii.copy_elemental_solution(sys, "teste", "e"); 
    } 
  } 
```

Reply:
I'll email you the working and failing versions of the MOOSE exodus output.

Reply:
But boy, I'd really like to have a proper test for this API, because if neither this nor my previous two attempts at writing unit tests for it worked, something's wrong.  "Roy's an idiot" can't be ruled out, I'm afraid, but I'd say "the API is too brittle" and "the API is too poorly documented" are at least in the running.

Reply:
I can replicate the error with this, thanks.

Reply:
I pushed a fix. The issue was that I was supposed to interpret an empty list of active subdomains as "active on all subdomains".

Also, @roystgnr, note that you need to add "exii.write_element_data(es);" after "exii.write_equation_systems("mesh_with_soln.e", es);" to get your test case to work.

Reply:
Ha!  I tried the one, I tried the other; I didn't try both.

Reply:
P.S. You also need to change "exii.write_equation_systems("mesh_with_soln.e", es);" to:

```
    std::set<std::string> sys_list;
    exii.write_equation_systems("mesh_with_soln.e", es, &sys_list);
```
otherwise you try to write out the element variables twice.

Reply:
Yes, exactly. It writes out a nodal version of "e" in your test case, so you have to skip that in order to not clash with the element version that we write out when we call write_element_data.

Reply:
Looks good to me, thanks!